### PR TITLE
Rulesets are loaded automatically and PlayMode is "extendable"

### PR DESCRIPTION
**Pull request by jorolf**
not a wrong remote this time ;)
I made this so people have an easier time making playmodes and sharing them

### CODE REVIEW

**Comment by peppy** [comment]
Here's a little hint: *never* ever *ever* write the same code in two different places. Find a way to avoid that at all costs.

File: osu.Desktop/Program.cs [hunk 1]
```diff
@@ -41,10 +41,9 @@ public static int Main(string[] args)
                 }
                 else
                 {
-                    Ruleset.Register(new OsuRuleset());
-                    Ruleset.Register(new TaikoRuleset());
-                    Ruleset.Register(new ManiaRuleset());
-                    Ruleset.Register(new CatchRuleset());
+                    if (Debugger.IsAttached)
+                        cwd = Directory.GetParent(Directory.GetParent(cwd).FullName).FullName;
```
Comment (peppy) [inline osu.Desktop/Program.cs]: no

File: osu.Game.Modes.Catch/CatchDifficultyCalculator.cs [hunk 2]
```diff
@@ -4,22 +4,21 @@
...
 using osu.Game.Modes.Objects;
-using System;
 using System.Collections.Generic;
 
 namespace osu.Game.Modes.Catch
 {
     public class CatchDifficultyCalculator : DifficultyCalculator<CatchBaseHit>
     {
-        protected override PlayMode PlayMode => PlayMode.Catch;
+        protected override int PlayMode => Modes.PlayMode.Catch;
```
Comment (peppy) [inline osu.Game.Modes.Catch/CatchDifficultyCalculator.cs]: you should not change this enum to an int. let it be an enum.

File: osu.Game/Modes/Ruleset.cs [hunk 3]
```diff
@@ -51,5 +58,28 @@ public static Ruleset GetRuleset(PlayMode mode)
...
+            foreach (string subDirectory in Directory.EnumerateDirectories(directory))
+                loadRulesets(subDirectory);
+
+            foreach (string file in Directory.EnumerateFiles(directory))
+            {
+                if (!file.EndsWith(".dll"))
+                    continue;
+                try
+                {
+                    var rulesets = Assembly.LoadFile(file).GetTypes().Where((Type t) => t.IsSubclassOf(typeof(Ruleset)));
```
Comment (peppy) [inline osu.Game/Modes/Ruleset.cs]: you are permanently loading every dll in the folder. the overhead is not only huge, but you are wasting memory with a potentially second copy of EVERY dll in memory.

File: osu.Game/Modes/Ruleset.cs [hunk 4]
```diff
@@ -51,5 +58,28 @@ public static Ruleset GetRuleset(PlayMode mode)
...
+                    foreach (Type rulesetType in rulesets)
+                    {
+                        Ruleset.Register(Activator.CreateInstance(rulesetType) as Ruleset);
+                    }
+
+                }
+                catch (Exception e) { }
```
Comment (peppy) [inline osu.Game/Modes/Ruleset.cs]: this is slack

Comment (Jorolf) [inline osu.Game/Modes/Ruleset.cs]: mscorlib.dll throws an exception when I try to load it
got a suggestion on how to change that?

**Comment by peppy** [comment]
Here are some suggestions:

- Squash commits locally that are "changed something" or "rearranged something". They shouldn't be separated unless they have meaning.
- Don't randomly clean up files unrelated to your change. You rearranged usings and touched random String->string files when they are completely unrelated to this change.

File: osu.Game/Modes/Ruleset.cs [hunk 5]
```diff
@@ -51,5 +58,35 @@ public static Ruleset GetRuleset(PlayMode mode)
 
             return Activator.CreateInstance(type) as Ruleset;
         }
+        
+        public static void LoadRulesetsFrom(string directory, string filter = "", AppDomain domain = null)
+        {
+            if (domain == null)
+            {
+                domain = AppDomain.CreateDomain("rulesetLoader");
+                LoadRulesetsFrom(directory, filter, domain);
```
Comment (peppy) [inline osu.Game/Modes/Ruleset.cs]: there is no reason to make this a recursive call.

File: osu.Game/Modes/Ruleset.cs [hunk 6]
```diff
@@ -51,5 +58,35 @@ public static Ruleset GetRuleset(PlayMode mode)
...
+                AppDomain.Unload(domain);
+            }
+            else
+            {
+                foreach (string file in Directory.EnumerateFiles(directory))
+                {
+                    try
+                    {
+                        if (!(file.Contains(filter)&&file.EndsWith(".dll")))
```
Comment (peppy) [inline osu.Game/Modes/Ruleset.cs]: && has spaces before and after

File: osu.Game/Modes/Ruleset.cs [hunk 7]
```diff
@@ -51,5 +58,35 @@ public static Ruleset GetRuleset(PlayMode mode)
...
+                            continue;
+                        
+                        var assembly = domain.Load(AssemblyName.GetAssemblyName(file));
```
Comment (peppy) [inline osu.Game/Modes/Ruleset.cs]: you're *still* loading every dll in the folder. this is NOT FEASIBLE

File: osu.Game/OsuGame.cs [hunk 8]
```diff
@@ -1,32 +1,32 @@
...
+using osu.Framework.Graphics.Transforms;
 using osu.Framework.Input;
-using osu.Game.Input;
-using OpenTK.Input;
 using osu.Framework.Logging;
-using osu.Game.Graphics.UserInterface.Volume;
-using osu.Game.Database;
-using osu.Framework.Allocation;
-using osu.Framework.Graphics.Transforms;
+using osu.Framework.Screens;
```
Comment (peppy) [inline osu.Game/OsuGame.cs]: ?

File: osu.Game/Overlays/MusicController.cs [hunk 9]
```diff
@@ -9,6 +9,7 @@
 using osu.Framework.Allocation;
 using osu.Framework.Audio.Track;
 using osu.Framework.Configuration;
+using osu.Framework.Extensions.Color4Extensions;
```
Comment (peppy) [inline osu.Game/Overlays/MusicController.cs]: ?

File: osu.Game/Overlays/Toolbar/Toolbar.cs [hunk 10]
```diff
@@ -10,8 +10,8 @@
 using osu.Framework.Graphics.Transforms;
 using osu.Framework.Input;
 using osu.Game.Graphics;
-using osu.Game.Modes;
 using OpenTK;
+using osu.Game.Modes;
```
Comment (peppy) [inline osu.Game/Overlays/Toolbar/Toolbar.cs]: ?

File: osu.Game/Modes/Ruleset.cs (hunk 7 above)
Comment (Jorolf) [inline osu.Game/Modes/Ruleset.cs]: well... I added a filter...?

**Comment by Jorolf** [comment]
new branch + splitting the 'parts' of this pr

**Pull request closed without merging**